### ISSUE TITLE
feature idea: user favorite song folder(s)

### ISSUE THREAD

**Issue by RodentVienna** [dguzek/Simply-Love-SM5#225]
Rhythm Lunatic posted a video on youtube one month ago, going into adding a favorites folder into any SM5 Theme:
https://www.youtube.com/watch?v=a8KNUGEZ5jE

is this a potential feature you guys want to looking into adding Simply Love for SM5?

some thoughts:
- compatibility check for 5.0.12, 5.1b and 5.3a releases
- since this is a user preference, where will it be stored in the user profile?
- how will it be stored (textfile with path-to-song? hashes from installed songs?)
- is it posible to have more than one favorites folder, and let the user decide on creating, naming and removing them?
- a potential fast way to 
- UX: player interaction to add/remove a song to a favorites folder (within sort menu, as item in modifier menu, input shortcut)
- iconize a favorized entry
- a way to export, share and import a favorites list, for sharing with friends
- handling orphaned entries, e.g. when using player profile on another machine, which has not installed all songs that are favorized
- in case this is relevant for machine options to be generally enabled/disabled

- even more things to consider

**Comment by RodentVienna** [comment]
while this looks feasible to be a theme-specific featue for Simply Love SM5, it may at some stage become part of the sm5.3-engine Team Rizu is working on. It may make sense to align with Team Rizu in this case.

**Comment by RodentVienna** [comment]
referencing this specific commit which got a favorites implementation started
(check the commits after this one as well), latest commit states that "cleanup is required")

https://github.com/CrashCringle12/Simply-Love-SM5/commit/776df02807bf87d9cf68dd41dc136f80232fa8e8

cannot judge quality, style and completeness (in terms of "does it cover use case, edge cases and compatibilities").
only serves as a reference point.

**Comment by CrashCringle12** [comment]
oh that's me

**Comment by CrashCringle12** [comment]
Yeah, that was the initial commit when I first went and tried to implement favorites. At this point we have it going pretty well and have been using it on our cab for a few years. I've tweaked the implementation up a bunch over time. There are some limitations to it tho which is why I never put in a PR for it but it works for our purposes.

I'll make a separate branch for specifically favorites to show u if you want to just cherry pick it or look at it. But the bulk of the methods are here: https://github.com/CrashCringle12/Simply-Love-SM5/blob/psu-cabby/Scripts/SL-FavoritesHandler.lua

**Comment by CrashCringle12** [comment]
(Addressing your thoughts) I can only answer based on our experience using the present system. I'm kind of haphazardly responding to this so if you want any additional clarification I can happily provide:

1 - It works for 5.0.12, 5.1b, and itgmania. I don't use outfox and I'm unsure how things are setup engine wise, so I'm not sure if it would work.

2. It stores files in the Profile's directory and will keep a temp file in the "Other" folder of the theme. SM has support for  Favorite-like systems via the Preferred Sort. It can read in a file x as long as it's been prefixed with "SongManager " and create groups based on the contents of the file. Here's an example of what the file looks like: https://github.com/CrashCringle12/Simply-Love-SM5/blob/psu-cabby/Other/SongManager%20FavoriteSongs.txt. 

3. Ye you can have multiple folders for it, folders are specified with the "--" prefix. Folders can't be created in gameplay, this could be implemented in theory tho, but I mostly play on a public cab so this wasn't something thought about. Ppl can create folders by modifying their favorites file in their profile directory. It's also worth noting that songs unlocked via the Unlock system automatically are included as a folder in the preferredsort if the preference is enabled.

4.  Orphaned entries are handled, if the song isn't there it just doesn't show up in the list. 
5. You can pretty easily just copy lines and entire files to share with each other.
6. I presently only allow favorites to be added on ScreenEvaluation, but this could be easily modified to allow on the music wheel. I only don't have enabled cause my fork is for our public cab, and our club members have like a bazillion favorites as it is. And I'm worried about someone adding all 18k songs to their favorites and issues happening.

The big limitation though is that SM doesn't like songs appearing in multiple sections/groups. The Preferred sort gets suuuper funky when the same song is referenced in multiple groups. I checked the engine and yeah that's unfortunately expected behavior.
